### PR TITLE
Removing insertion of llvm.emit_c_interface attribute

### PR DESCRIPTION
**Pull request by nhasabni**
This removal is needed primarily because 1) we do not need this attribute, 2) IREE's ConvertToLLVM pass breaks because of this attribute, and 3) more importantly, this attribute generates unnecessary overhead in LLVM IR while invoking XSMM/perf functions (see below).

This commit also cleans up duplicate code in TPP runtime for XsmmRunnerUtils.

The commit addresses issue #463.

LLVM IR with the attribute:
---------------------------

Notice below how `xsmm_matmul_invoke` entry point available in `XsmmRunnerUtils` is called. With the `emit_c_interface` attribute, MLIR generates a stub function - `xsmm_matmul_invoke` - to call into `XsmmRunnerUtils`, the inputs of which (e.g., `%9`) are repacked from `%arg6` and `%arg7`. `xsmm_matmul_invoke` then invokes `_mlir_ciface_xsmm_matmul_invoke`, which is defined in `XsmmRunnerUtils`.

In summary, `_entry -> xsmm_matmul_invoke -> _mlir_ciface_xsmm_matmul_invoke (XsmmRunnerUtils)`.

```
llvm.func _entry(…) {
  …
  %43 = llvm.alloca %42 x !llvm.struct<(ptr, ptr, i64, array<2 x i64>, array<2 x i64>)> : (i64) -> !llvm.ptr
  llvm.store %23, %43 : !llvm.struct<(ptr, ptr, i64, array<2 x i64>, array<2 x i64>)>, !llvm.ptr
  %44 = llvm.mlir.constant(2 : index) : i64
  %45 = llvm.mlir.undef : !llvm.struct<(i64, ptr)>
  %46 = llvm.insertvalue %44, %45[0] : !llvm.struct<(i64, ptr)>
  %47 = llvm.insertvalue %43, %46[1] : !llvm.struct<(i64, ptr)>
  llvm.call @xsmm_matmul_invoke(…, %44, %43) : (i64, i64, i64, !llvm.ptr, i64,
                                                !llvm.ptr, i64, !llvm.ptr) -> ()
  llvm.return
}

llvm.func @xsmm_matmul_invoke(…, %arg6: i64, %arg7: !llvm.ptr) attributes
        {llvm.emit_c_interface, sym_visibility = "private"} {
  …
  %9 = llvm.alloca %8 x !llvm.struct<(i64, ptr)> : (i64) -> !llvm.ptr
  llvm.store %7, %9 : !llvm.struct<(i64, ptr)>, !llvm.ptr
  %10 = llvm.mlir.undef : !llvm.struct<(i64, ptr)>
  %11 = llvm.insertvalue %arg6, %10[0] : !llvm.struct<(i64, ptr)>
  %12 = llvm.insertvalue %arg7, %11[1] : !llvm.struct<(i64, ptr)>
  %13 = llvm.mlir.constant(1 : index) : i64
  %14 = llvm.alloca %13 x !llvm.struct<(i64, ptr)> : (i64) -> !llvm.ptr
  llvm.store %12, %14 : !llvm.struct<(i64, ptr)>, !llvm.ptr
  llvm.call @_mlir_ciface_xsmm_matmul_invoke(%arg0, %arg1, %4, %9, %14) :
                                            (i64, i64, !llvm.ptr, !llvm.ptr, !llvm.ptr) -> ()
  llvm.return
}
```

LLVM IR without the attribute:
------------------------------

Without the attribute, the stub function is not generated. Instead, `_entry` calls into `xsmm_matmul_invoke` which is defined in our `XsmmRunnerUtils`.

In summary, `_entry -> xsmm_matmul_invoke (XsmmRunnerUtils)`.

```
// This function is same as before.
llvm.func _entry(…) {
  …
  %43 = llvm.alloca %42 x !llvm.struct<(ptr, ptr, i64, array<2 x i64>, array<2 x i64>)> : (i64) -> !llvm.ptr
  llvm.store %23, %43 : !llvm.struct<(ptr, ptr, i64, array<2 x i64>, array<2 x i64>)>, !llvm.ptr
  %44 = llvm.mlir.constant(2 : index) : i64
  %45 = llvm.mlir.undef : !llvm.struct<(i64, ptr)>
  %46 = llvm.insertvalue %44, %45[0] : !llvm.struct<(i64, ptr)>
  %47 = llvm.insertvalue %43, %46[1] : !llvm.struct<(i64, ptr)>
  llvm.call @xsmm_matmul_invoke(…, %44, %43) : (i64, i64, i64, !llvm.ptr, i64,
                                                !llvm.ptr, i64, !llvm.ptr) -> ()
  llvm.return
}

// This is just a declaration. Definition is found in XsmmRunnerUtils. 
llvm.func @xsmm_matmul_invoke(i64, i64, i64, !llvm.ptr, i64, !llvm.ptr, i64, !llvm.ptr) attributes {sym_visibility = "private"}
```

### CODE REVIEW

**Comment by rengolin** [comment]
I believe we used this attribute to avoid fiddling with void pointers, but if it breaks IREE and generates useless stubs, it may be worth having the extra code in our runtime (which seems small enough).

the perf runtime also uses the same attribute, perhaps we should common up the utility functions in a separate file and get rid of the attribute everywhere.

@chelini , how does this affect the `useMeta` flag?

**Review by chelini — COMMENTED** [review]
Thanks.

File: runtime/XsmmRunnerUtils.cpp [hunk 1]
```diff
@@ -25,69 +25,58 @@ static void printXsmmStruct(const libxsmm_meltw_binary_shape &binaryShape,
...
+  }
+  fprintf(stderr, "Unhandled data type in get_data_pointer_from_memref_desc:%d",
+          dType);
+  return nullptr;
+}
+} // namespace
 
+extern "C"
+void xsmm_matmul_invoke(const libxsmm_datatype dType, int64_t addr,
+                        int64_t rankA, void *memrefDescA,
```
[the remark this answers is not in the document]
Comment (nhasabni) [reply runtime/XsmmRunnerUtils.cpp:60]: That the default approach of LLVM/MLIR to pass a memref to an external function - tuple of `rank, memref_descriptor`.

[the remark this answers is not in the document]
Comment (chelini) [reply runtime/XsmmRunnerUtils.cpp:60]: I thought the default was to just splat all the memref fields in the function signature.

[the remark this answers is not in the document]
Comment (chelini) [reply runtime/XsmmRunnerUtils.cpp:60]: Indeed:
```
func.func @main(%arg0: memref<4x4xf32>) {
  return
}
```
```
module attributes {llvm.data_layout = ""} {
  llvm.func @main(%arg0: !llvm.ptr, %arg1: !llvm.ptr, %arg2: i64, %arg3: i64, %arg4: i64, %arg5: i64, %arg6: i64) {
    llvm.return
  }
}
```
with `./tpp-opt func.mlir -convert-func-to-llvm`

**Comment by rengolin** [comment]
What's the status here? This PR is trailing and collecting conflicts.

**Comment by rengolin** [comment]
Do we have a consensus? What's the best approach to have in the runtime library?

**Comment by chelini** [comment]
FYI: https://github.com/llvm/llvm-project/commit/e02d4142dca48d9359ad304fec629ea3d7e6924c

**Comment by chelini** [comment]
> Do we have a consensus? What's the best approach to have in the runtime library?

We should align with IREE to avoid duplicating the runtime. But there is something that needs to be understood better. For example, I was expecting to be able to remove these casts (https://github.com/plaidml/tpp-mlir/blob/85aceb35c5306a73960810951b9636eff74aa5c9/lib/TPP/ConvertXsmmToFunc.cpp#L38) as now we explicitly pass the rank in the function signature, but it seems not the case.

**Comment by chelini** [comment]
@nhasabni I am happy with this update. Please have a look. If you merge please do not squash and preserve the commit train.

**Comment by rengolin** [comment]
> @nhasabni I am happy with this update. Please have a look. If you merge please do not squash and preserve the commit train.

Please, only rebase merge if you have no merge commits. Otherwise, please rebase on your local branch and push again to keep linear history.

**Comment by nhasabni** [comment]
> > @nhasabni I am happy with this update. Please have a look. If you merge please do not squash and preserve the commit train.
> 
> Please, only rebase merge if you have no merge commits. Otherwise, please rebase on your local branch and push again to keep linear history.

I just used GitHub's option through UI. I will do it through editor next time.

**Comment by rengolin** [comment]
> I just used GitHub's option through UI. I will do it through editor next time.

Github has an option to do the same thing as a rebase, it's a drop down on the same button. :)

**Comment by nhasabni** [comment]
@rengolin @adam-smnk Can one of you please review this PR? Thanks.

**Review by rengolin — APPROVED** [review]

**Comment by rengolin** [comment]
May be nothing, but we got a 12% regression in one benchmark:
```
matmul_256x1024x4096_fp32_mlir
matmul_fp32_omp_16_mlir | 1472.957 | 1297.942 | 88%
```

**Comment by alheinecke** [comment]
is this reproducible across several runs / sizes

**Comment by rengolin** [comment]
> is this reproducible across several runs / sizes

Looking at historical values, just wanted to dump here so we don't lose track.

Only that one size/type was affected this much (all others were the same with some +-4% at most), it may have been a glitch in the node.

**Comment by rengolin** [comment]
Seems to be high variation of benchmarks:
| Run | GFLOP/s |
| -- | -- |
| 47 | 1298.884 |
| 55 | 1491.453 |
| 61 | 1472.957 |
| 62 | 1297.942 |

Looks bi-modal, could be node-specific.

**Comment by alheinecke** [comment]
are they running turbo or not? Turbo spread is 10-15%

**Comment by rengolin** [comment]
> are they running turbo or not? Turbo spread is 10-15%

Yes! Ok, that explains it then.

**Pull request merged**